### PR TITLE
support texture binding when bindless is not available

### PR DESCRIPTION
**Pull request by ruancomelli**
## Summary by CodeRabbit

- **New Features**
	- Introduced command-line argument handling for enhanced application configurability.
	- Added support for controlling bindless texture functionality.
	- Improved texture management with dynamic shader assignment based on texture state.

- **Bug Fixes**
	- Enhanced error handling for texture initialization and rendering processes to prevent null reference issues.

- **Documentation**
	- Updated comments and debug prints to clarify new command-line options and texture management logic.

### CODE REVIEW

**Comment by ruancomelli** [comment]
This is an experiment review for experiment `multiple_mermaid_diagram_types`
Run ID: `multiple_mermaid_diagram_types/run_2024-09-26T16-03-48_v1-23-0-21-g3eb85d2d2`
The benchmark review for this pull request can be found at https://github.com/sourcery-ai-experiments/foundations/pull/13.
This pull request was cloned from `https://github.com/btipling/foundations/pull/33`. _(Note: the URL is not a link to avoid triggering a notification on the original pull request.)_

<details>
<summary>Experiment configuration</summary>

```yaml
review_config:
  # User configuration for the review
  # - benchmark - use the user config from the benchmark reviews
  # - <value> - use the value directly
  user_review_config:
    enable_ai_review: true
    enable_rule_comments: false

    enable_complexity_comments: false
    enable_security_comments: false
    enable_tests_comments: false
    enable_comment_suggestions: false
    enable_functionality_review: false

    enable_pull_request_summary: false
    enable_review_guide: true

    enable_approvals: false

  ai_review_config:
    # The model responses to use for the experiment
    # - benchmark - use the model responses from the benchmark reviews
    # - llm - call the language model to generate responses
    model_responses:
      comments_model: benchmark
      comment_area_model: benchmark
      comment_validation_model: benchmark
      comment_suggestion_model: benchmark
      complexity_model: benchmark
      functionality_model: benchmark
      security_model: benchmark
      tests_model: benchmark
      pull_request_summary_model: benchmark
      review_guide_model: llm
      overall_comments_model: benchmark

# The pull request dataset to run the experiment on
pull_request_dataset:
- https://github.com/sourcery-ai/core/pull/4607
- https://github.com/sourcery-ai/core/pull/4631
- https://github.com/sourcery-ai/core/pull/4647
    # CodeRabbit examples:
- https://github.com/2lambda123/-Orange-OpenSource-oorobot/pull/15
- https://github.com/2lambda123/galaxyproject-galaxy/pull/12
- https://github.com/a0v0/avtoolz/pull/79
- https://github.com/adityask98/Hotaru/pull/10
- https://github.com/agdas/vscode/pull/2
- https://github.com/agluszak/hirschgarten/pull/2
- https://github.com/alexsnow348/insightface/pull/46
- https://github.com/alikuxac/utilities/pull/10
- https://github.com/AlphaDev87/timba-api/pull/49
- https://github.com/AngeloTadeucci/Maple2/pull/239
- https://github.com/AngeloTadeucci/Maple2.File/pull/36
- https://github.com/AngeloTadeucci/Maple2/pull/233
    # Examples where CodeRabbit does not generate diagrams
- https://github.com/baptisteArno/typebot.io/pull/1778
- https://github.com/btipling/foundations/pull/33
- https://github.com/btipling/foundations/pull/31
- https://github.com/chintu-777/jaeger/pull/1
- https://github.com/coji/remix-docs-ja/pull/55
- https://github.com/DaveMBush/SmartNgRX/pull/622
- https://github.com/DaveMBush/SmartNgRX/pull/481
- https://github.com/dkittle/party-connections/pull/6
- https://github.com/Drajad-Kusuma-Adi/onstudy-backend/pull/6
- https://github.com/imaami/libcanth/pull/2
    # Requested by Tim
- https://github.com/2lambda123/-Orange-OpenSource-oorobot/pull/15
- https://github.com/2lambda123/galaxyproject-galaxy/pull/12
- https://github.com/a0v0/avtoolz/pull/79
- https://github.com/adityask98/Hotaru/pull/10
- https://github.com/agdas/vscode/pull/2
- https://github.com/agluszak/hirschgarten/pull/2
- https://github.com/4DNucleome/PartSeg/pull/1183
- https://github.com/ably/ably-cocoa/pull/1973
- https://github.com/ably-labs/ably-chat-swift/pull/54
- https://github.com/ably-labs/ably-chat-swift/pull/35
  # Including database changes to evaluate entity-relationship diagrams
- https://github.com/sourcery-ai/core/pull/4579
  # TODO: find some with Alembic

# Questions to ask to label the review comments
review_comment_labels: []
# - label: correct
#   question: Is this comment correct?

# Benchmark reviews generated by running
#   python -m scripts.experiment benchmark <experiment_name>
benchmark_reviews:
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4607
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/374
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4631
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/375
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4647
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/376
- dataset_pull_request: https://github.com/2lambda123/-Orange-OpenSource-oorobot/pull/15
  review_pull_request: https://github.com/sourcery-ai-experiments/-Orange-OpenSource-oorobot/pull/5
- dataset_pull_request: https://github.com/2lambda123/galaxyproject-galaxy/pull/12
  review_pull_request: https://github.com/sourcery-ai-experiments/galaxyproject-galaxy/pull/13
- dataset_pull_request: https://github.com/adityask98/Hotaru/pull/10
  review_pull_request: https://github.com/sourcery-ai-experiments/Hotaru/pull/14
- dataset_pull_request: https://github.com/agdas/vscode/pull/2
  review_pull_request: https://github.com/sourcery-ai-experiments/vscode/pull/14
- dataset_pull_request: https://github.com/agluszak/hirschgarten/pull/2
  review_pull_request: https://github.com/sourcery-ai-experiments/hirschgarten/pull/13
- dataset_pull_request: https://github.com/alikuxac/utilities/pull/10
  review_pull_request: https://github.com/sourcery-ai-experiments/utilities/pull/4
- dataset_pull_request: https://github.com/AlphaDev87/timba-api/pull/49
  review_pull_request: https://github.com/sourcery-ai-experiments/timba-api/pull/12
- dataset_pull_request: https://github.com/AngeloTadeucci/Maple2/pull/239
  review_pull_request: https://github.com/sourcery-ai-experiments/Maple2/pull/23
- dataset_pull_request: https://github.com/AngeloTadeucci/Maple2.File/pull/36
  review_pull_request: https://github.com/sourcery-ai-experiments/Maple2.File/pull/12
- dataset_pull_request: https://github.com/AngeloTadeucci/Maple2/pull/233
  review_pull_request: https://github.com/sourcery-ai-experiments/Maple2/pull/24
- dataset_pull_request: https://github.com/baptisteArno/typebot.io/pull/1778
  review_pull_request: https://github.com/sourcery-ai-experiments/typebot.io/pull/7
- dataset_pull_request: https://github.com/btipling/foundations/pull/33
  review_pull_request: https://github.com/sourcery-ai-experiments/foundations/pull/13
- dataset_pull_request: https://github.com/btipling/foundations/pull/31
  review_pull_request: https://github.com/sourcery-ai-experiments/foundations/pull/14
- dataset_pull_request: https://github.com/chintu-777/jaeger/pull/1
  review_pull_request: https://github.com/sourcery-ai-experiments/jaeger/pull/7
- dataset_pull_request: https://github.com/coji/remix-docs-ja/pull/55
  review_pull_request: https://github.com/sourcery-ai-experiments/remix-docs-ja/pull/7
- dataset_pull_request: https://github.com/DaveMBush/SmartNgRX/pull/622
  review_pull_request: https://github.com/sourcery-ai-experiments/SmartNgRX/pull/13
- dataset_pull_request: https://github.com/DaveMBush/SmartNgRX/pull/481
  review_pull_request: https://github.com/sourcery-ai-experiments/SmartNgRX/pull/14
- dataset_pull_request: https://github.com/dkittle/party-connections/pull/6
  review_pull_request: https://github.com/sourcery-ai-experiments/party-connections/pull/7
- dataset_pull_request: https://github.com/Drajad-Kusuma-Adi/onstudy-backend/pull/6
  review_pull_request: https://github.com/sourcery-ai-experiments/onstudy-backend/pull/7
- dataset_pull_request: https://github.com/imaami/libcanth/pull/2
  review_pull_request: https://github.com/sourcery-ai-experiments/libcanth/pull/7
- dataset_pull_request: https://github.com/2lambda123/-Orange-OpenSource-oorobot/pull/15
  review_pull_request: https://github.com/sourcery-ai-experiments/-Orange-OpenSource-oorobot/pull/6
- dataset_pull_request: https://github.com/2lambda123/galaxyproject-galaxy/pull/12
  review_pull_request: https://github.com/sourcery-ai-experiments/galaxyproject-galaxy/pull/14
- dataset_pull_request: https://github.com/adityask98/Hotaru/pull/10
  review_pull_request: https://github.com/sourcery-ai-experiments/Hotaru/pull/15
- dataset_pull_request: https://github.com/agdas/vscode/pull/2
  review_pull_request: https://github.com/sourcery-ai-experiments/vscode/pull/15
- dataset_pull_request: https://github.com/agluszak/hirschgarten/pull/2
  review_pull_request: https://github.com/sourcery-ai-experiments/hirschgarten/pull/14
- dataset_pull_request: https://github.com/4DNucleome/PartSeg/pull/1183
  review_pull_request: https://github.com/sourcery-ai-experiments/PartSeg/pull/7
- dataset_pull_request: https://github.com/ably/ably-cocoa/pull/1973
  review_pull_request: https://github.com/sourcery-ai-experiments/ably-cocoa/pull/6
- dataset_pull_request: https://github.com/ably-labs/ably-chat-swift/pull/54
  review_pull_request: https://github.com/sourcery-ai-experiments/ably-chat-swift/pull/11
- dataset_pull_request: https://github.com/ably-labs/ably-chat-swift/pull/35
  review_pull_request: https://github.com/sourcery-ai-experiments/ably-chat-swift/pull/12
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4579
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/377

```

</details>

**Comment by SourceryAI** [comment]
## Reviewer's Guide by Sourcery

This pull request adds support for texture binding when bindless textures are not available. It introduces a new command-line argument to disable bindless textures, modifies the texture initialization and binding process, and updates several scene files to accommodate these changes.

### Sequence Diagram

```mermaid
sequenceDiagram
    participant App
    participant Args
    participant Texture
    participant OpenGL
    App->>Args: init()
    Args-->>App: args
    App->>Texture: init(args.disable_bindless)
    Texture->>OpenGL: Check GL_ARB_bindless_texture support
    alt Bindless supported and not disabled
        Texture->>OpenGL: Create bindless handle
        Texture->>OpenGL: Make texture resident
    else Bindless not supported or disabled
        Texture->>OpenGL: Use standard texture binding
    end
    App->>Texture: bind()
    alt Bindless enabled
        Texture->>OpenGL: Set uniform handle
    else Bindless disabled
        Texture->>OpenGL: Bind texture unit
    end
```

### File-Level Changes

| Change | Details | Files |
| ------ | ------- | ----- |
| Added support for non-bindless texture binding | <ul><li>Introduced a 'disable_bindless' flag in the Texture struct</li><li>Modified the Texture initialization process to handle both bindless and non-bindless cases</li><li>Added a fallback mechanism to create a magenta texture when image data is not available</li><li>Updated the texture binding process to use glBindTextureUnit when bindless is disabled</li></ul> | `src/foundations/rhi/Texture.zig` |
| Implemented command-line argument handling | <ul><li>Created a new Args struct to handle command-line arguments</li><li>Added a '--disable_bindless' flag to control bindless texture functionality</li><li>Integrated the Args struct into the application initialization process</li></ul> | `src/foundations/Args.zig`<br/>`src/foundations/app.zig` |
| Updated scene files to support the new texture binding system | <ul><li>Modified texture initialization in various scene files to use the new Texture.init method</li><li>Updated fragment shader selection based on the bindless texture state</li><li>Improved error handling for texture loading and initialization</li></ul> | `src/foundations/scenes/cgpoc/chapter5/textured_pyramid/TexturedPyramid.zig`<br/>`src/foundations/scenes/cgpoc/chapter6/textured_torus/TexturedTorus.zig`<br/>`src/foundations/scenes/cgpoc/chapter6/earth/Earth.zig`<br/>`src/foundations/scenes/cgpoc/chapter6/shuttle/Shuttle.zig` |

---

<details>
<summary>Tips and commands</summary>

#### Interacting with Sourcery

- **Trigger a new review:** Comment `@sourcery-ai review` on the pull request.
- **Continue discussions:** Reply directly to Sourcery's review comments.
- **Generate a GitHub issue from a review comment:** Ask Sourcery to create an
  issue from a review comment by replying to it.

#### Customizing Your Experience

Access your [dashboard](https://app.sourcery.ai) to:
- Enable or disable review features such as the Sourcery-generated pull request
  summary, the reviewer's guide, and others.
- Change the review language.
- Add, remove or edit custom review instructions.
- Adjust other review settings.

#### Getting Help

- [Contact our support team](mailto:support@sourcery.ai) for questions or feedback.
- Visit our [documentation](https://docs.sourcery.ai) for detailed guides and information.
- Keep in touch with the Sourcery team by following us on [X/Twitter](https://x.com/SourceryAI), [LinkedIn](https://www.linkedin.com/company/sourceryai/) or [GitHub](https://github.com/sourcery-ai).

</details>

File: src/foundations/rhi/Texture.zig [hunk 1]
```diff
@@ -4,45 +4,68 @@ handle: c.GLuint64 = 0,
...
+    };
+
     if (c.glfwExtensionSupported("GL_ARB_bindless_texture") != 1 or c.glfwExtensionSupported("GL_ARB_bindless_texture") != 1) {
-        return TextureError.BindlessNotSupported;
+        t.disable_bindless = true;
     }
+    return t;
+}
 
+pub fn setup(self: *Texture, image: ?*assets.Image, program: u32, uniform_name: []const u8) TextureError!void {
```
Comment (SourceryAI) [inline src/foundations/rhi/Texture.zig:31]: **suggestion (bug_risk):** Consider more robust error handling in texture setup

The current implementation might silently fail in some error cases. Consider propagating errors or logging them for better debugging and reliability.

File: src/foundations/scenes/cgpoc/chapter5/textured_pyramid/TexturedPyramid.zig [hunk 2]
```diff
@@ -88,11 +88,13 @@ pub fn updateCamera(_: *TexturedPyramid) void {}
 
 pub fn renderParallepiped(self: *TexturedPyramid) void {
     const prog = rhi.createProgram();
+    self.brick_texture = rhi.Texture.init(self.ctx.args.disable_bindless) catch null;
```
Comment (SourceryAI) [inline src/foundations/scenes/cgpoc/chapter5/textured_pyramid/TexturedPyramid.zig:91]: **suggestion (performance):** Evaluate performance impact of disabling bindless textures

While the flexibility is good, consider documenting or measuring the performance difference between bindless and non-bindless textures to guide users on when to use this option.

File: src/foundations/rhi/Texture.zig [hunk 3]
```diff
@@ -4,45 +4,68 @@ handle: c.GLuint64 = 0,
...
-        0,
-        0,
-        0,
-        @intCast(image.width),
-        @intCast(image.height),
-        c.GL_RGBA,
-        c.GL_UNSIGNED_BYTE,
-        image.stb_data.ptr,
-    );
+    if (image) |img| {
```
Comment (SourceryAI) [inline src/foundations/rhi/Texture.zig:38]: **suggestion:** Consider handling texture creation failure

The code creates a magenta texture when no image is provided, but doesn't handle the case where texture creation fails. Consider adding error handling for this scenario.

**Review by SourceryAI — COMMENTED** [review]
Hey @ruancomelli - I've reviewed your changes and they look great!

<details>
<summary>Here's what I looked at during the review</summary>

- 🟡 **General issues**: 3 issues found
- 🟢 **Documentation**: all looks good
</details>

***

[LangSmith trace](https://smith.langchain.com/o/7d8518ad-6dd3-5c7b-96d2-757769296335/projects/p/88560225-5072-48bc-967c-dc54da53f265/r/1eca1ef9-f9db-4612-b19f-ccf220a888bc?poll=true)

<sub>
Help me be more useful! Please click 👍 or 👎 on each comment to tell me if it was helpful.
</sub>

**Pull request closed without merging**